core/src/main/java/org/infinispan/util/DoubleIterator.java

```diff
@@ -0,0 +1,74 @@
+package org.infinispan.util;
+
+import java.util.NoSuchElementException;
+import java.util.function.Supplier;
+
+import org.infinispan.commons.util.CloseableIterator;
+
+/**
+ * Iterator implementation that will return all entries from the first iterator. Upon completion of the first iterator
+ * the supplier will generate an additional iterator that will then be used as a source for this iterator. After the
+ * second iterator is consumed this iterator will also be completed.
+ * @param <E> element type
+ */
+public class DoubleIterator<E> implements CloseableIterator<E> {
+   private final CloseableIterator<E> iterator1;
+   private final Supplier<? extends CloseableIterator<E>> supplier;
+
+   private CloseableIterator<E> iterator2;
+   private E iterator2Next;
+
+   public DoubleIterator(CloseableIterator<E> first, Supplier<? extends CloseableIterator<E>> supplier) {
+      this.iterator1 = first;
+      this.supplier = supplier;
+   }
+
+   @Override
+   public void close() {
+      try (CloseableIterator<E> closeme = iterator1) {
+         if (iterator2 != null) {
+            iterator2.close();
+         }
+      }
+   }
+
+   @Override
+   public boolean hasNext() {
+      boolean hasNext;
+      if (iterator2 == null) {
+         hasNext = iterator1.hasNext();
+         if (hasNext) {
+            return true;
+         } else {
+            iterator2 = supplier.get();
+         }
+      }
+
+      if (iterator2.hasNext()) {
+         iterator2Next = iterator2.next();
+      }
+      return iterator2Next != null;
```

Comment:
Hrmm I agree about null. To be honest this class was based off of a previous one that required objects to be non null. However there is no reason this one can't allow null values. I can fix this.

Actually tbh with you I think the whole iteration process will have lots of trouble returning null values (reactive stream spec for example does not allow null values). Normally you can't get a null value from a ConcurrentMap, but unfortunately if you use a stream with a map operator you can get an iterator that returns null there.

I am not sure if we want to start supporting null all the way down the stack or not. It will require most likely a lot of changes.